### ISSUE TITLE
Add notebook execution buttons to the sphinx gallery notebook pages

### ISSUE THREAD

**Issue by romanlutz** [fairlearn/fairlearn#764]
Our notebook pages on the website (such as [this one](https://fairlearn.org/v0.6.1/auto_examples/plot_grid_search_census.html)) have download buttons, but we can't execute them directly on Binder, Azure Notebooks, Google Colab, etc.

Scikit-learn has a button for Binder (e.g., in [this example](https://scikit-learn.org/stable/auto_examples/bicluster/plot_spectral_biclustering.html#sphx-glr-auto-examples-bicluster-plot-spectral-biclustering-py)) so this should be possible to add since we're using a similar set of sphinx extensions such as sphinx-gallery.

The way scikit-learn achieves this is by adding the following lines to the sphinx gallery configuration: https://github.com/scikit-learn/scikit-learn/blob/309f135c3284d7db6e23ca81a87948c7066a3949/doc/conf.py#L338
... and there's a dedicated `requirements.txt` file.

Other than adding that requirements file and the configuration for sphinx gallery, this would require building the website. Instructions are on [this page](https://fairlearn.org/main/contributor_guide/development_process.html#building-the-website), but I suggest going with the "simple" method rather than building the website for all versions. After all, we wouldn't expect the button to show up in past versions.

**Comment by kurianbenoy** [comment]
Hi I would like to work on this issue.

**Comment by romanlutz** [comment]
I expect the config values needed will be
```
'binder': {
        'org': 'fairlearn',
        'repo': 'fairlearn', 
        'binderhub_url': 'https://mybinder.org', 
        'branch': binder_branch, 
        'dependencies': './binder/requirements.txt', 
        'use_jupyter_lab': True    
    }
```

As you can tell the branch changes depending on the version since the documentation is built in multiple versions. I'm hoping the same version parsing will work for us (https://github.com/scikit-learn/scikit-learn/blob/309f135c3284d7db6e23ca81a87948c7066a3949/doc/conf.py#L282 and following), but it's not clear that it will. You may have to try it out. If that doesn't work, just use "main" as the branch. The dependencies file is a new file (which is mentioned in the issue description). I have no particular opinion on jupyter lab, but you can try it with `False` and `True` and see if it works.

**Comment by kurianbenoy** [comment]
Thanks @romanlutz for your quick reply. Is the requirements.txt used in `dependencies` pointing to a folder `binder/requirements.txt` in our case?

If you look  at scikit docs, [the requirements.txt](https://github.com/scikit-learn/scikit-learn/blob/309f135c3284d7db6e23ca81a87948c7066a3949/doc/binder/requirements.txt) requirements.txt there is actually nothing. So what should be there in requirements.txt in our case? 

Also it would be very helpful if you can point to some link which you used to setup sphinx_gallery configuration for our project, to get a better idea of the library used.

**Comment by romanlutz** [comment]
Seems like scikit-learn hs a special setup with a separate repo for this. You can check their binder repo to see what's in the requirements file. Probably just the main dependencies we have in our requirements.txt plus the package itself. I'd just try it out by pointing at an empty file and if it fils try adding dependencies. Binder is free so there's no real cost associated with trying this.

You can check the Sphinx Gallery docs https://sphinx-gallery.github.io/stable/index.html
It's a very cool tool with lots of functionality that we're not even using right now.

**Comment by kurianbenoy** [comment]
@romanlutz, was a bit busy the past few weeks. I couldn't work on it for a few weeks now, but I would definitely like to continue doing this if there is no time pressure.

**Comment by romanlutz** [comment]
@kurianbenoy there is no pressure at all! Good to know, and let us know if there's any way we can help.

**Comment by kurianbenoy** [comment]
@romanlutz how can I bring the docs website up?

https://fairlearn.org/main/contributor_guide/contributing_documentation.html. Is it by the command:

`make doc`

**Comment by romanlutz** [comment]
https://fairlearn.org/main/contributor_guide/contributing_documentation.html should help 🙂 

I don't have a preference between lab and notebook. How does it affect the task?

**Comment by kurianbenoy** [comment]
Yeah it doesn't affect the task in hand. One issue, I am facing now is when passing dependencies in binder config, I am getting the below error

![image](https://user-images.githubusercontent.com/24592806/122648028-d23b1200-d144-11eb-8f99-fef3f3abd409.png)

**Comment by kurianbenoy** [comment]
Also another thing is I am able to get the run binder button: 
![image](https://user-images.githubusercontent.com/24592806/122648229-d87dbe00-d145-11eb-8c33-59f43b5f0b9b.png)

Yet on selecting it, I am getting this error

![image](https://user-images.githubusercontent.com/24592806/122648328-71acd480-d146-11eb-9dc9-bcb568bf7133.png)

I will be trying to debug these issue for a couple of hours more and see if I can get a solution. If any help is available it would be great as well :)

**Comment by kurianbenoy** [comment]
@romanlutz just noticed when checking our docs/conf.py that the copyright is hardcoded. Can we make it dynamic like the Scikit-learn docs configuration?

https://github.com/scikit-learn/scikit-learn/blob/309f135c3284d7db6e23ca81a87948c7066a3949/doc/conf.py#L88

**Comment by kurianbenoy** [comment]
I am a bit confused on what should we give for the `notebook_dir` parameter in binder configuration. In case of `scikit-learn` and `sphinx-gallery` configuration, they don't have any folder named notebooks in their repo. But that is not the case of fairlearn which has a notebooks folder. 

Reference link:
https://sphinx-gallery.github.io/stable/configuration.html#binder-links

**Comment by adrinjalali** [comment]
maybe this would be of some help? https://github.com/scikit-learn/scikit-learn/tree/main/.binder

**Comment by kurianbenoy** [comment]
I will check it out. Maybe the postbuild file maybe helpful to solve the issue.

**Comment by romanlutz** [comment]
A few clarifications:
- Ignore the `notebooks` directory. These are only notebooks that existed before we started using `sphinx-gallery`. Eventually they'll either be migrated to the gallery format (basically using `.py` files with the magic markers to separate cells) or deleted.
- Examples/notebooks to render on the webpage are in the `examples` directory. So you can use whatever scikit-learn specifies for notebook directory (which may just be nothing in case they use a default). This should fix the path issue for `plot_quickstart_selection_rate.py`
- Copyright: I don't particularly care to be honest, but if you like feel free to make the year dynamic. I guess the start year would be 2018.
- In the issue description I referred to the `requirements.txt` file from scikit-learn. Did you add one for Fairlearn? That might explain the `binder/requirements.txt` issue.

**Comment by kurianbenoy** [comment]
Ok, I am ignoring the notebooks section and using default value for notebooks. Also I have added requirements.txt in .binder folder.
